compiler/backend/src/org/jetbrains/kotlin/codegen/inline/GlobalInlineContext.kt

```diff
@@ -5,24 +5,44 @@
 
 package org.jetbrains.kotlin.codegen.inline
 
-import org.jetbrains.kotlin.codegen.InlineCycleReporter
+import com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.descriptors.CallableDescriptor
 import org.jetbrains.kotlin.diagnostics.DiagnosticSink
-import org.jetbrains.kotlin.resolve.calls.model.ResolvedCall
+import org.jetbrains.kotlin.diagnostics.Errors
 import java.util.*
 
-class GlobalInlineContext(diagnostics: DiagnosticSink) {
-
-    private val inlineCycleReporter: InlineCycleReporter = InlineCycleReporter(diagnostics)
+class GlobalInlineContext(private val diagnostics: DiagnosticSink) {
+    private val inlineCalls = linkedMapOf<CallableDescriptor, PsiElement?>()
 
     private val typesUsedInInlineFunctions = LinkedList<MutableSet<String>>()
 
-    fun enterIntoInlining(call: InlineCall?) =
-        inlineCycleReporter.enterIntoInlining(call).also {
-            if (it) typesUsedInInlineFunctions.push(hashSetOf())
+    fun enterDeclaration(descriptor: CallableDescriptor) {
+        assert(descriptor.original !in inlineCalls) { "entered inlining cycle on $descriptor" }
+        inlineCalls[descriptor.original] = null
+    }
+
+    fun exitDeclaration(descriptor: CallableDescriptor) {
+        inlineCalls.remove(descriptor.original)
+    }
+
+    fun enterIntoInlining(context: CallableDescriptor, callee: CallableDescriptor?, element: PsiElement?): Boolean {
+        assert(context.original in inlineCalls) { "$context not in current declaration stack" }
+        inlineCalls[context.original] = element
+        if (callee?.original in inlineCalls) {
+            var currentElement = element
+            for ((callContainer, callElement) in inlineCalls.entries.dropWhile { it.key != callee?.original }) {
+                currentElement?.let { diagnostics.report(Errors.INLINE_CALL_CYCLE.on(it, callContainer)) }
+                currentElement = callElement // next container is the target of this call
+            }
+            inlineCalls[context.original] = null
+            return false
         }
+        typesUsedInInlineFunctions.push(hashSetOf())
+        return true
+    }
 
-    fun exitFromInliningOf(call: InlineCall?) {
-        inlineCycleReporter.exitFromInliningOf(call)
+    fun exitFromInlining(context: CallableDescriptor) {
+        inlineCalls[context.original] = null
```

Comment:
Shoudn't it be 'remove'?

Reply:
The original idea was that enterDeclaration adds the element, then enterIntoInlining sets the value, then exitFromInlining clears it, then steps 2-3 are repeated, and finally exitDeclaration removes the key. That's not relevant anymore.